### DISCUSSION TITLE
Type checking for curried/partially applied functions

### DISCUSSION THREAD

**Discussion by molotch** [microsoft/pyright#1950] [Ideas]
It would be very nice to have static type checking for curried and/or partially applied functions if possible.

I'm thinking about the partial function from the functools module or the @curry decorator from the Pymonad package.

**Comment by erictraut** [comment]
I'm not aware of anything in the current Python type standards that would support this. PEP 612 (ParamSpec) allows limited forms of parameter concatenation and removal, but you're probably looking for something more general.

If you have a specific proposal that you'd like to suggest, I recommend that you post it in the Python [typing-sig](https://mail.python.org/archives/list/typing-sig@python.org/). That's where new Python type system standards are discussed.

**Reply by erictraut** [reply]
We could consider adding special-case knowledge of `functools.partial` and hard-coding its behavior in the type checker. We generally avoid such special-case hackery because it's not a scalable approach, but we have made exceptions in a few cases where there is sufficient interest. We can use this discussion to track upvotes.

We wouldn't consider adding special-case knowledge of functions in third-party libraries like Pymonad, so this wouldn't fully address your needs.

**Reply by molotch** [reply]
Thank you for the feedback, I have no idea about the Python type standards. Just thought it would improve the experience of writing Python in a functional way. As of now I'm probably gonna rewrite functions to return inner functions instead of using partial just to get type hint support.

Lets see if anyone else upvotes.
